### ISSUE TITLE
Add outOfStoreSymlink to `home.file`

### ISSUE THREAD

**Issue by DrRuhe** [nix-community/home-manager#3514]
I'd like to have an configuration option for `home.file` entries, that allows to declare these files as outOfStoreSymlinks.
So something like this:
```nix
home.file.".config" = {
   source = ./config;
   recursive = true;
   outOfStoreSymlink = true;
};
```
This would allow me to have a config folder for the related things in my dotfiles under VCS, but still allow the files to be changed by the programs. Changes could then be managed by the VCS.

As a concrete example: [The config files of KDE and related programs are plenty](https://github.com/shalva97/kde-configuration-files). And I'd like to manage these via nix and track them in the dotfiles. When these link to the files in the dotfiles directory, I could still use the GUI for managing the Settings, which would not be possible with links to immutable files in the nix store. 

In particular, I'd like it to work with the recursive option mentioned above since AFAICT currently it is only possible to do this with single files.

**Comment by montchr** [comment]
In my experience with a flakes-only configuration, I misused `mkOutOfStoreSymlink` initially, expecting that a relative path (like `./config` in your example) would not result in an out-of-store symlink at all, because a relative path in a flake is always relative to the flake's store path. It took some time to figure out that it only has the desired effect when its argument is an absolute path on the filesystem. AFAIK, there's no "flake-native" way to detect the flake's out-of-store path on a filesystem, so I've resorted to a concoction of custom module options, environment variables, and hard-coded paths to get the desired behavior.

So, with that said, I suspect that adding an option to the file linker options as described would also necessitate the addition of _at least_ one other option to define an assumed base path for the flake-based home-manager configuration. 

But overall I agree that it would be nice to establish some clearer conventions for dealing with this kind of setup!

**Comment by montchr** [comment]
I marked my previous comment as "outdated" since I'm not sure if my musings were accurate or not. TL;DR I've been under the impression that `mkOutOfStoreSymlink` requires an absolute path argument, but the (old) discussion in #257 seems to indicate otherwise, so perhaps I need to update my understanding of that function again 🙃

**Comment by eliandoran** [comment]
@montchr , I had a similar experience on my side.

Using the following:
```nix
{
  home.file = {
    ".config/hypr/hyprland.conf".source = config.lib.file.mkOutOfStoreSymlink ./hyprland/hyprland.conf;
  };
}
```

Results in the creation of the following symlink chain:

```
~/.config/hypr/hyprland.conf
 -> /nix/store/k76965c6lx8xdqc2swha74s8k0hw7xgc-home-manager-files/.config/hypr/hyprland.conf
 -> /nix/store/9wm464xzpmwnaav4q0li0nmdm98slfab-hm_hyprland.conf
 -> /nix/store/mnaq3pzm0qrkvs8kpgpjw456iprj36sy-source/core/hyprland/hyprland.conf
```

In the previous case, the symlink failed since it was pointing to a missing file.

Passing a string instead, the path will be relative to the Nix store:
```
/nix/store/5rnm8ch7ybg5ddk1jj20c2avrbrji0af-hm_hyprland.conf -> ./hyprland/hyprland.conf
```

Thus, an absolute path must be used.

**Comment by montchr** [comment]
A potential benefit of a dedicated option for OOS-symlinks could be validating the type of the option definition. Perhaps a warning could be logged if the value points to a store path, which is most likely not desirable.

**Comment by pearsche** [comment]
i just needed this so this is relevant still

**Comment by sealor** [comment]
For me, this issue is still relevant too!

I'm also using a flake for my configuration. The following helper function solved my issues with the absolute path:
```
createSymlink = localPath: config.lib.file.mkOutOfStoreSymlink "${config.xdg.configHome}/home-manager/${localPath}";
```

In my case I would like to use:
```
home.file."bin" = {
   source = "${config.xdg.configHome}/home-manager/bin";
   recursive = true;
   outOfStoreSymlink = true;
};
```

It would be nice to have this feature in future versions. Thanks!

**Comment by tjkirch** [comment]
Not stale.

**Comment by MarvinJWendt** [comment]
This is still very relevant. A popular linux YouTuber just uploaded a video why he is switching from home-manager to stow again (https://www.youtube.com/watch?v=U6reJVR3FfA). This issue is the exact reason.

We need a way to make configurations editable without rebuilding. If you have a big neovim setup via nix, you tend to change the configuration multiple times a day.

I personally struggle with adding `nushell` scripts to my config (dotfiles), as always need to rebuild the system, before I can actually run my scripts. This makes development very slow.

I cannot use an absolute path, as my configurations have different base paths on NixOS and macOS.

**Comment by hakan-demirli** [comment]
I agree that the issue is still relevant and it would be a great QoL improvement. But, the problem in the video is the evaluation time of the home-manager/nix and the fact that he does not need reproducibility and rollback features of the nix. So, it is logical to go back to stow. Why pay for something you are not using. However, he can use any one of the workarounds listed in this thread to symlink his nvim config.

As for your abs path problem, there are couple of workarounds.

You can pass different configs to mac and nixos.
Pseudo Code:
```nix
    let
      systemSettings = {
        system = "x86_64-linux";
      };

      userSettingsShared =  {
        username = "myname"; 
      };
      userSettingsNixOS =  {
        dotfilesDir = "/home/${userSettingsShared.username}/Desktop/dotfiles";
      };
      userSettingsMacOS =  {
        dotfilesDir = "whatever mac does for paths";
      };
    in
```
```nix
      homeConfigurations."${userSettingsShared.username}_macos" = home-manager.lib.homeManagerConfiguration {
        pkgs = import nixpkgs {
          system = systemSettings.system;
        };
        extraSpecialArgs = {
          inherit inputs userSettingsSharedMacOS userSettingsShared;
        };
```
```nix
      homeConfigurations."${userSettingsShared.username}_nixos" = home-manager.lib.homeManagerConfiguration {
        pkgs = import nixpkgs {
          system = systemSettings.system;
        };
        extraSpecialArgs = {
          inherit inputs userSettingsSharedNixOS userSettingsShared;
        };
```
Then activate the one you choose with: `home-manager switch --flake ~/Desktop/dotfiles/#myname_macos`

**Comment by amardeep** [comment]
This is the workaround I am using - https://github.com/nix-community/home-manager/issues/2085#issuecomment-2022239332

**Comment by codethief** [comment]
Not stale

**Comment by zaporter** [comment]
Not stale

**Comment by NoCodeNoBlunder** [comment]
Any progress on this?

**Comment by ardling** [comment]
Please add a feature to create symbolic links using Home Manager. It would be useful in any case, since symbolic links often appear in the home directory. It’s especially convenient when setting up my environment, for example, to create a link from one file to another, maybe binary.
